src/TestCase.ts

```diff
@@ -0,0 +1,173 @@
+import * as yaml from "js-yaml";
+import * as vscode from "vscode";
+import { Position, Range, Selection } from "vscode";
+import NavigationMap from "./NavigationMap";
+import { ActionType, PartialTarget, PrimitiveTarget, Target } from "./Types";
+
+export type SerializedPosition = {
+  line: number;
+  character: number;
+};
+
+export type SerializedRange = {
+  start: SerializedPosition;
+  end: SerializedPosition;
+};
+
+export type SerializedSelection = {
+  anchor: SerializedPosition;
+  active: SerializedPosition;
+};
+
+export function serializeRange(range: Range): SerializedRange {
+  return {
+    start: serializePosition(range.start),
+    end: serializePosition(range.end),
+  };
+}
+
+export function serializeSelection(selection: Selection): SerializedSelection {
+  return {
+    active: serializePosition(selection.active),
+    anchor: serializePosition(selection.anchor),
+  };
+}
+
+export function serializePosition(position: Position): SerializedPosition {
+  return { line: position.line, character: position.character };
+}
+
+type Command = {
+  actionName: ActionType;
+  partialTargets: PartialTarget[];
+  extraArgs: any[];
+};
+
+type Snapshot = {
+  document: string;
+  clipboard: string;
+  visibleRanges: SerializedRange[];
+  selections: SerializedSelection[];
+};
+
+type DecorationRanges = { [coloredSymbol: string]: SerializedRange };
+
+export type TestCaseFixture = {
+  command: Command;
+  languageId: string;
+  decorations: DecorationRanges;
+  initialState: Snapshot;
+  finalState: Snapshot;
+};
+
+export default class TestCase {
+  command: Command;
+  languageId: string;
+  decorations: DecorationRanges;
+  initialState: Snapshot | null = null;
+  finalState: Snapshot | null = null;
+
+  constructor(
+    command: Command,
+    targets: Target[],
+    navigationMap: NavigationMap
+  ) {
+    const activeEditor = vscode.window.activeTextEditor!;
+
+    this.command = command;
+    this.languageId = activeEditor.document.languageId;
+    this.decorations = this.extractTargetedDecorations(targets, navigationMap);
+  }
+
+  extractPrimitiveTargetKeys(...targets: PrimitiveTarget[]) {
+    const keys: string[] = [];
+    targets.forEach((target) => {
+      if (target.mark.type === "decoratedSymbol") {
+        keys.push(`${target.mark.symbolColor}.${target.mark.character}`);
```

Comment:
I tried to use the function in navigation map for this so that we don't duplicate the key generation here